File: lib/aoc24/day05.clj
(ns aoc24.day05
  (:require [core :as c]
            [clojure.string :as s]))

(defn parse-input [file]
  (let [text (slurp file)
        [rules pages] (s/split text #"\n\n")
        rules (->> (s/split rules #"\n")
                   (map #(s/split % #"\|"))
                   (map reverse)
                   (map #(mapv parse-long %))
                   (group-by first)
                   (map (fn [[a b]] [a (set (map second b))]))
                   (into {}))
        pages (->> (s/split pages #"\n")
                   (map #(s/split % #","))
                   (map #(map parse-long %)))]
    [rules pages]))

(defn ordered-page [page rules]
  (sort-by #(count (filter (get rules % #{}) page)) page))

(defn middle-of-page-num [page]
  (nth page (quot (count page) 2)))

(defn part1 [file]
  (let [[rules pages] (parse-input file)]
    (->> pages
         (filter #(= (ordered-page % rules) %))
         (map middle-of-page-num)
         (c/sum))))

(defn part2 [file]
  (let [[rules pages] (parse-input file)]
    (->> pages
         (keep (fn [page]
                 (->> (ordered-page page rules)
                      (c/then [ordered-page]
                              (when (not= page ordered-page)
                                (middle-of-page-num ordered-page))))))
         (c/sum))))

File: lib/aoc24/day05.ex
defmodule Aoc24.Day05 do
  def part1(file) do
    {rules, pages} = rules_and_pages(file)

    pages
    |> Enum.filter(&(&1 == ordered_page(&1, rules)))
    |> Enum.map(&middle/1)
    |> Enum.sum()
  end

  def part2(file) do
    {rules, pages} = rules_and_pages(file)

    pages
    |> Enum.map(&ordered_page(&1, rules))
    |> Enum.zip(pages)
    |> Enum.filter(fn {ordered, unordered} -> ordered != unordered end)
    |> Enum.map(fn {ordered, _} -> ordered end)
    |> Enum.map(&middle/1)
    |> Enum.sum()
  end

  defp middle(coll) do
    middle = coll |> length() |> div(2)

    Enum.at(coll, middle)
  end

  defp ordered_page(page, rules) do
    Enum.sort_by(page, fn itm ->
      before = Map.get(rules, itm, [])

      Enum.count(page, &(&1 in before))
    end)
  end

  def rules_and_pages(file) do
    [rules, pages] =
      file
      |> File.read!()
      |> String.split("\n\n")

    {rules(rules), pages(pages)}
  end

  defp rules(rules) do
    rules
    |> String.trim()
    |> String.split("\n")
    |> Enum.map(&String.split(&1, "|"))
    |> Enum.map(&Enum.reverse/1)
    |> Enum.map(fn nums -> Enum.map(nums, &String.to_integer/1) end)
    |> Enum.group_by(&List.first/1)
    |> Enum.map(fn {key, nums} -> {key, Enum.map(nums, fn [_, i] -> i end)} end)
    |> Map.new()
  end

  defp pages(pages) do
    pages
    |> String.trim()
    |> String.split("\n")
    |> Enum.map(&String.split(&1, ","))
    |> Enum.map(fn nums -> Enum.map(nums, &String.to_integer/1) end)
  end
end
